### ISSUE TITLE
SASA based force

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#4746]
Implicit solvent models often include a term that is proportional to the solvent accessible surface area (SASA).  It's challenging to compute, especially since you need the derivatives as well.  GBSAOBCForce and AmoebaGeneralizedKirkwoodForce both use a fast but inaccurate approximation.

I want to add a force class that computes it accurately rather than just approximately.  The state of the art for computing it is based on alpha shape theory.  Here are some papers describing implementations of it on CPUs and GPUs, respectively.

https://doi.org/10.1021/acs.jcim.2c01346
https://doi.org/10.1021/acs.jctc.3c01366

Both those implementations are open source (LGPL for AlphaMol, MIT for dSASA), so we should be able to either use them directly or adapt them.

**Comment by epretti** [comment]
We've discussed that there are challenges with an efficient implementation of the exact method.  I've taken a brief look into approximate methods that will hopefully be more accurate than what OpenMM currently supports.  It should definitely be possible to implement [LCPO](https://doi.org/10.1002/(SICI)1096-987X(19990130)20:2%3C217::AID-JCC4%3E3.0.CO;2-A), which is basically a particular form of three-body potential.  Since the method's parameters are assigned per atom based on elements and bond topology, the potential could be implemented in the library layer and the parameters assigned in the application layer (we'd have to think about the best way to expose this).

We could also consider [pwSASA](https://doi.org/10.1021/acs.jctc.8b00413), a newer method that Amber supports (`gbsa=3`).  In graduate school, a coworker and I actually implemented this for OpenMM; it's just a pairwise potential that we implemented with `CustomNonbondedForce`, and I don't know whether we'd need a more complicated implementation.  The most annoying part was hand-translating a hardcoded table from Fortran to Python for looking up the parameters, which Amber of course decided to do based on residue names.

**Comment by epretti** [comment]
To follow up on this with a bit more information and thoughts about the API: at one point we had mentioned having something like a `SASAForce` for this.  However, it may be better to have separate classes for different methods, since the underlying functional forms as well as necessary parameters are different:

- The existing ACE model is probably well-suited for staying inside `GBSAOBCForce` since it is dependent on the computed Born radii.
- LCPO:
  - Requires specifying a radius and coefficients $P_1$, $P_2$, $P_3$, $P_4$ for each atom (type).
  - In Amber's implementation, parameter assignment is done by some Fortran logic (if you have a copy of sander source code, see `mdread2.F90`).  It should be possible to replicate this exactly when reading from a prmtop.  It does rely on looking up Amber atom type names at some places, so I'd have to think about how to support this from the ForceField route.
- pwSASA (if we want it):
  - Requires specifying an epsilon, sigma, cutoff, and maximum SASA value for each atom (type).
  - The pair potential part of this could theoretically be implemented in the Python application layer with `CustomNonbondedForce`.  However, the formulation also includes a constant energy offset corresponding to all atoms being separated from each other (where the SASA is the sum of the atom sphere surface areas).  This doesn't affect forces for MD, but it should probably be included, so maybe it'd be better to have a force class to compute this whole thing.
  - Parameter assignment in Amber's implementation similarly uses Fortran logic (see `gbsa3_init` in sander `egb.F90`).  Again, if reading from a prmtop, it should be possible to replicate the logic.  Again, I'm not sure about the best way to do this with ForceField.  Relying on atom type names isn't necessarily enough because of the way the pwSASA parameters were fit; some lookups to residue template names are required.
- If we try to implement the accurate (alpha shape theory) method at some point, that could probably have its own force class since the implementation will be very different from the other methods.

**Comment by peastman** [comment]
I'm not familiar with pwSASA.  From the paper, it's not entirely clear what the benefits are.  The main benefit they describe is that it runs on the GPU in Amber, making it much faster.  But it's not clear whether that's because LCPO can't be efficiently implemented on a GPU, or just that no one had gotten around to doing it.

I also checked the Amber24 manual, which has this to say about it:

> Although gbsa = 3 is supported in pmemd, its general usage is not recommended as the
speed gain is insignificant because its fast approximation of surface area algorithm was specifically
designed for GPU-accelerated GB simulations. Therefore, we recommend users use gbsa=3 with
pmemd.cuda.

So basically they recommend LCPO on the CPU and pwSASA on the GPU.  But if they have similar accuracy I'd prefer to support only one.  The question is whether they can be implemented efficiently on both CPU and GPU.

**Comment by epretti** [comment]
An implementation of LCPO using `CustomManyParticleForce` can be found at <https://gist.github.com/epretti/1578a498e0841907c50c7f90eff5e9bd/archive/089df9c2d4f2ffa03d5aaeef45e225f27d815939.zip>.  This includes a test input system and configuration (DHFR in implicit solvent, 2489 atoms).  See lines 54-96 for the `CustomManyParticleForce` functional forms and options.  The code below this is just logic for assigning parameters from a lookup table.

I suspect that implementing and optimizing an `LCPOForce` will be faster, but you can take a look at what is here and see if there is anything obvious that would be causing a major slowdown.  The largest single chunk of time is taken computing the 3-body interactions:

<img width="1436" height="481" alt="Image" src="https://github.com/user-attachments/assets/fc34e71f-f143-4e42-bee3-6f4571b0cd19" />

**Comment by peastman** [comment]
I don't see any obvious improvements to the way you've written the CustomManyParticleForce.  If we want to try to optimize CustomManyParticleForce itself, there's one change that would make a lot of difference for this case.

You want to ignore all hydrogens when computing the surface area.  You do that with type filters, requiring all particles have type 1.  That isn't a situation I was thinking about when I implemented CustomManyParticleForce.  I was thinking of cases where all particles would be used, but some would be treated differently from others.  In this case, it could realize that particles of type 0 will never be used at all, and therefore filter them out right from the beginning.  They wouldn't even appear in the neighbor list.  Assuming half the particles are hydrogen, it would only have to consider 1/8 as many interactions.

I don't know if this situation is common in other cases.  For this case, LCPO is probably special enough (and important enough) that we want an optimized implementation of it anyway.
